File: app/src/main/java/me/echeung/moemoekyun/api/service/AuthService.java
package me.echeung.moemoekyun.api.service;

import lombok.AllArgsConstructor;
import me.echeung.moemoekyun.api.ErrorHandlingAdapter;
import me.echeung.moemoekyun.api.response.AuthResponse;
import me.echeung.moemoekyun.api.response.BaseResponse;
import retrofit2.http.Body;
import retrofit2.http.Header;
import retrofit2.http.POST;

public interface AuthService {
    @POST("register")
    ErrorHandlingAdapter.WrappedCall<BaseResponse> register(@Body RegisterBody body);

    @POST("login")
    ErrorHandlingAdapter.WrappedCall<AuthResponse> login(@Body LoginBody body);

    @POST("login/mfa")
    ErrorHandlingAdapter.WrappedCall<AuthResponse> mfa(@Header("Authorization") String token, @Body LoginMfaBody body);

    @AllArgsConstructor
    class RegisterBody {
        String email;
        String username;
        String password;
    }

    @AllArgsConstructor
    class LoginBody {
        String username;
        String password;
    }

    @AllArgsConstructor
    class LoginMfaBody {
        String token;
    }
}

File: app/src/main/java/me/echeung/moemoekyun/client/api/service/AuthService.kt
package me.echeung.moemoekyun.client.api.service

import me.echeung.moemoekyun.client.api.ErrorHandlingAdapter
import me.echeung.moemoekyun.client.api.response.AuthResponse
import me.echeung.moemoekyun.client.api.response.BaseResponse
import retrofit2.http.Body
import retrofit2.http.Header
import retrofit2.http.POST

interface AuthService {
    @POST("register")
    fun register(@Body body: RegisterBody): ErrorHandlingAdapter.WrappedCall<BaseResponse>

    @POST("login")
    fun login(@Body body: LoginBody): ErrorHandlingAdapter.WrappedCall<AuthResponse>

    @POST("login/mfa")
    fun mfa(@Header("Authorization") token: String, @Body body: LoginMfaBody): ErrorHandlingAdapter.WrappedCall<AuthResponse>

    class RegisterBody(val email: String, val username: String, val password: String)

    class LoginBody(val username: String, val password: String)

    class LoginMfaBody(val token: String)
}
